File: cli/src/documentation_robotics/core/projection_engine.py
"""
Projection engine - automatically creates elements across layers.
"""

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Optional

import yaml
from jinja2 import Template


@dataclass
class ProjectionRule:
    """Defines how to project from one layer to another."""

    name: str
    from_layer: str
    from_type: str
    to_layer: str
    to_type: str
    name_template: str
    property_mappings: Dict[str, str]
    conditions: Optional[Dict[str, Any]] = None
    template_file: Optional[str] = None


class ProjectionEngine:
    """
    Manages cross-layer projections.

    Projects elements from source layers to target layers based on rules.
    """

    def __init__(self, model: Any, rules_path: Optional[Path] = None):
        """
        Initialize projection engine.

        Args:
            model: The architecture model
            rules_path: Path to projection-rules.yaml
        """
        self.model = model
        self.rules: List[ProjectionRule] = []

        if rules_path and rules_path.exists():
            self.load_rules(rules_path)

    def load_rules(self, path: Path) -> None:
        """
        Load projection rules from file.

        Args:
            path: Path to projection-rules.yaml
        """
        with open(path, "r") as f:
            data = yaml.safe_load(f)

        if not data or "projections" not in data:
            return

        for rule_data in data["projections"]:
            self.rules.append(self._parse_rule(rule_data))

    def _parse_rule(self, rule_data: Dict[str, Any]) -> ProjectionRule:
        """Parse a projection rule from YAML data."""
        # Parse from/to specifications
        from_spec = rule_data["from"].split(".")
        from_layer = from_spec[0]
        from_type = from_spec[1] if len(from_spec) > 1 else None

        to_spec = rule_data["to"].split(".")
        to_layer = to_spec[0]
        to_type = to_spec[1] if len(to_spec) > 1 else None

        # Get first rule (simplified - could support multiple)
        rule_details = rule_data["rules"][0] if "rules" in rule_data else {}

        return ProjectionRule(
            name=rule_data.get("name", f"{from_layer}-to-{to_layer}"),
            from_layer=from_layer,
            from_type=from_type,
            to_layer=to_layer,
            to_type=rule_details.get("create_type", to_type),
            name_template=rule_details.get("name_template", "{source.name}"),
            property_mappings=rule_details.get("properties", {}),
            conditions=rule_data.get("conditions"),
            template_file=rule_details.get("template"),
        )

    def find_applicable_rules(
        self, source: Any, target_layer: Optional[str] = None
    ) -> List[ProjectionRule]:
        """
        Find projection rules applicable to a source element.

        Args:
            source: Source element
            target_layer: Optional target layer filter

        Returns:
            List of applicable projection rules
        """
        applicable = []

        for rule in self.rules:
            # Check layer match
            if rule.from_layer != source.layer:
                continue

            # Check type match
            if rule.from_type and rule.from_type != source.type:
                continue

            # Check target layer filter
            if target_layer and rule.to_layer != target_layer:
                continue

            # Check conditions
            if rule.conditions and not self._check_conditions(source, rule.conditions):
                continue

            applicable.append(rule)

        return applicable

    def _check_conditions(self, element: Any, conditions: Dict[str, Any]) -> bool:
        """Check if element meets projection conditions."""
        for key, expected_value in conditions.items():
            actual_value = element.get(key)

            if isinstance(expected_value, list):
                if actual_value not in expected_value:
                    return False
            elif actual_value != expected_value:
                return False

        return True

    def project_element(
        self,
        source: Any,
        target_layer: str,
        rule: Optional[ProjectionRule] = None,
        dry_run: bool = False,
    ) -> Optional[Any]:
        """
        Project an element to a target layer.

        Args:
            source: Source element to project
            target_layer: Target layer name
            rule: Optional specific rule to use
            dry_run: If True, don't actually create the element

        Returns:
            Projected element if successful, None otherwise
        """
        # Find rule if not provided
        if not rule:
            rules = self.find_applicable_rules(source, target_layer)
            if not rules:
                raise ValueError(f"No projection rule found for {source.id} -> {target_layer}")
            rule = rules[0]  # Use first matching rule

        # Build projected element
        projected = self._build_projected_element(source, rule)

        if dry_run:
            return projected

        # Add to model
        try:
            self.model.add_element(rule.to_layer, projected)

            # Update reference registry
            if hasattr(self.model, "reference_registry"):
                self.model.reference_registry.register_element(projected)

            return projected

        except Exception as e:
            raise ValueError(f"Failed to project element: {e}")

    def _build_projected_element(self, source: Any, rule: ProjectionRule) -> Any:
        """
        Build projected element from source and rule.

        Args:
            source: Source element
            rule: Projection rule

        Returns:
            New projected element
        """
        from ..utils.id_generator import generate_element_id
        from .element import Element

        # Render name template
        name = self._render_template(rule.name_template, source)

        # Generate ID
        element_id = generate_element_id(rule.to_layer, rule.to_type, name)

        # Build data dictionary
        data = {
            "id": element_id,
            "name": name,
        }

        # Add mapped properties
        for target_prop, source_template in rule.property_mappings.items():
            value = self._render_template(source_template, source)
            self._set_nested_property(data, target_prop, value)

        # Create element
        return Element(id=element_id, element_type=rule.to_type, layer=rule.to_layer, data=data)

    def _render_template(self, template_str: str, source: Any) -> str:
        """
        Render a template string with source element data.

        Args:
            template_str: Template string (e.g., "{source.name}Service")
            source: Source element

        Returns:
            Rendered string
        """
        from ..utils.id_generator import to_kebab_case

        # Create template context with nested structure for Jinja2
        source_dict = {
            **source.data,
            "id": source.id,
            "layer": source.layer,
            "type": source.type,
            "name": source.name,
            "name_pascal": self._to_pascal_case(source.name),
            "name_kebab": to_kebab_case(source.name),
            "name_snake": self._to_snake_case(source.name),
        }

        context = {"source": source_dict}

        # Render template (support both Jinja2 and simple format)
        if "{{" in template_str:
            # Jinja2 style - supports nested access
            template = Template(template_str)
            return template.render(**context)
        else:
            # Python format style - use string replacement
            # Replace {source.xxx} with actual values
            result = template_str
            for key, value in source_dict.items():
                placeholder = f"{{source.{key}}}"
                if placeholder in result:
                    result = result.replace(placeholder, str(value))

            return result

    def _to_pascal_case(self, text: str) -> str:
        """Convert to PascalCase."""
        words = text.replace("-", " ").replace("_", " ").split()
        return "".join(word.capitalize() for word in words)

    def _to_snake_case(self, text: str) -> str:
        """Convert to snake_case."""
        from ..utils.id_generator import to_kebab_case

        return to_kebab_case(text).replace("-", "_")

    def _set_nested_property(self, data: dict, path: str, value: Any) -> None:
        """Set a nested property using dot notation."""
        keys = path.split(".")
        current = data

        for key in keys[:-1]:
            if key not in current:
                current[key] = {}
            current = current[key]

        current[keys[-1]] = value

    def project_all(
        self,
        from_layer: Optional[str] = None,
        to_layer: Optional[str] = None,
        dry_run: bool = False,
    ) -> List[Any]:
        """
        Project all applicable elements.

        Args:
            from_layer: Optional source layer filter
            to_layer: Optional target layer filter
            dry_run: If True, don't actually create elements

        Returns:
            List of projected elements
        """
        projected = []

        # Get source elements
        source_elements = []

        if from_layer:
            layer = self.model.get_layer(from_layer)
            if layer:
                source_elements = list(layer.elements.values())
        else:
            for layer in self.model.layers.values():
                source_elements.extend(layer.elements.values())

        # Project each element
        for source in source_elements:
            rules = self.find_applicable_rules(source, to_layer)

            for rule in rules:
                try:
                    element = self.project_element(source, rule.to_layer, rule, dry_run)
                    if element:
                        projected.append(element)
                except Exception as e:
                    # Log error but continue
                    print(f"Warning: Failed to project {source.id}: {e}")

        return projected

File: cli/src/core/projection-engine.ts
/**
 * Projection Engine - Python CLI Compatible Implementation
 *
 * Automatically creates elements across layers based on projection rules
 * with property transformations, conditional logic, and template rendering.
 *
 * Spec: cli-validation/projection-engine-spec.md
 */

import { readFile } from "fs/promises";
import { parse as parseYAML } from "yaml";
import { Model } from "./model.js";
import { Element } from "./element.js";
import { ReferenceRegistry } from "./reference-registry.js";
import { getErrorMessage } from "../utils/errors.js";

/**
 * Transform types for property mapping
 */
export type TransformType =
  | "uppercase"
  | "lowercase"
  | "kebab"
  | "snake"
  | "pascal"
  | "prefix"
  | "suffix"
  | "template";

/**
 * Condition operators for filtering
 */
export type ConditionOperator =
  | "exists"
  | "equals"
  | "not_equals"
  | "contains"
  | "matches"
  | "gt"
  | "lt"
  | "in";

/**
 * Property transformation configuration
 */
export interface PropertyTransform {
  type: TransformType;
  value?: string; // For prefix/suffix/template
}

/**
 * Condition for filtering source elements
 */
export interface ProjectionCondition {
  field: string;
  operator: ConditionOperator;
  value?: unknown;
  pattern?: string; // For regex matching
}

/**
 * Property mapping from source to target
 */
export interface PropertyMapping {
  source: string;
  target: string;
  default?: unknown;
  required?: boolean;
  transform?: PropertyTransform;
}

/**
 * Projection rule definition
 */
export interface ProjectionRule {
  name: string;
  from_layer: string;
  from_type: string;
  to_layer: string;
  to_type: string;
  name_template: string;
  property_mappings: PropertyMapping[];
  conditions?: ProjectionCondition[];
  template_file?: string;
  create_bidirectional?: boolean;
}

/**
 * Helper class for property transformations
 */
export class PropertyTransformer {
  constructor(private config: PropertyTransform) {}

  apply(value: unknown): string | null {
    if (value === null || value === undefined) return null;

    const strValue = String(value);

    switch (this.config.type) {
      case "uppercase":
        return strValue.toUpperCase();

      case "lowercase":
        return strValue.toLowerCase();

      case "kebab":
        return toKebabCase(strValue);

      case "snake":
        return toSnakeCase(strValue);

      case "pascal":
        return toPascalCase(strValue);

      case "prefix":
        return `${this.config.value}${strValue}`;

      case "suffix":
        return `${strValue}${this.config.value}`;

      case "template":
        return this.config.value?.replace("{value}", strValue) ?? strValue;

      default:
        return strValue;
    }
  }
}

/**
 * Helper class for condition evaluation
 */
export class ConditionEvaluator {
  constructor(private config: ProjectionCondition) {}

  evaluate(element: unknown): boolean {
    const fieldValue = getNestedProperty(element as Record<string, unknown>, this.config.field);

    switch (this.config.operator) {
      case "exists":
        return fieldValue !== null && fieldValue !== undefined;

      case "equals":
        return fieldValue === this.config.value;

      case "not_equals":
        return fieldValue !== this.config.value;

      case "contains":
        return fieldValue ? String(fieldValue).includes(String(this.config.value)) : false;

      case "matches":
        return fieldValue && this.config.pattern
          ? new RegExp(this.config.pattern).test(String(fieldValue))
          : false;

      case "gt":
        return (fieldValue as number) > (this.config.value as number);

      case "lt":
        return (fieldValue as number) < (this.config.value as number);

      case "in":
        return Array.isArray(this.config.value) ? this.config.value.includes(fieldValue) : false;

      default:
        return false;
    }
  }
}

/**
 * Projection Engine - Creates elements across layers based on rules
 */
export class ProjectionEngine {
  private rules: ProjectionRule[] = [];
  private registry?: ReferenceRegistry;

  constructor(
    private model: Model,
    rulesPath?: string
  ) {
    if (rulesPath) {
      this.loadRulesSync(rulesPath);
    }
  }

  /**
   * Set reference registry for element registration
   */
  setRegistry(registry: ReferenceRegistry): void {
    this.registry = registry;
  }

  /**
   * Load projection rules from YAML file
   */
  async loadRules(path: string): Promise<void> {
    try {
      const content = await readFile(path, "utf-8");
      const data = parseYAML(content) as Record<string, unknown>;

      if (!data.projections || !Array.isArray(data.projections)) {
        throw new Error("Invalid projection rules file: missing projections array");
      }

      this.rules = (data.projections as Record<string, unknown>[]).map((proj) =>
        this.parseProjectionRule(proj)
      );
    } catch (error) {
      throw new Error(
        `Failed to load projection rules from ${path}: ${getErrorMessage(error)}`
      );
    }
  }

  /**
   * Synchronous version for constructor
   */
  private loadRulesSync(_path: string): void {
    // For now, rules must be loaded async via loadRules()
    // This is a placeholder for backward compatibility
  }

  /**
   * Add a projection rule programmatically
   */
  addRule(rule: ProjectionRule): void {
    this.rules.push(rule);
  }

  /**
   * Get all loaded rules
   */
  getRules(): ProjectionRule[] {
    return [...this.rules];
  }

  /**
   * Clear all rules
   */
  clearRules(): void {
    this.rules = [];
  }

  /**
   * Parse YAML projection definition into ProjectionRule
   */
  private parseProjectionRule(data: Record<string, unknown>): ProjectionRule {
    // Split from/to into layer.type
    const [from_layer, from_type] = (data.from as string).split(".");
    const [to_layer, to_type] = (data.to as string).split(".");

    // Get first rule definition (simplified - Python CLI could support multiple)
    const ruleData = Array.isArray(data.rules)
      ? (data.rules as Record<string, unknown>[])[0]
      : (data.rules as Record<string, unknown>);

    // Parse conditions
    const conditions: ProjectionCondition[] = (
      (data.conditions as Record<string, unknown>[] | undefined) || []
    ).map((c: Record<string, unknown>) => ({
      field: c.field as string,
      operator: c.operator as ConditionOperator,
      value: c.value,
      pattern: c.pattern as string | undefined,
    }));

    // Parse property mappings
    const property_mappings: PropertyMapping[] = [];
    if (
      ruleData &&
      typeof ruleData === "object" &&
      "properties" in ruleData &&
      ruleData.properties
    ) {
      const properties = ruleData.properties as Record<string, unknown>;
      for (const [target, mapping] of Object.entries(properties)) {
        if (typeof mapping === "string") {
          // Simple format: { target: source }
          property_mappings.push({ source: mapping, target });
        } else if (typeof mapping === "object" && mapping !== null) {
          // Advanced format with transform
          const m = mapping as Record<string, unknown>;
          property_mappings.push({
            source: m.source as string,
            target,
            default: m.default,
            required: (m.required as boolean | undefined) ?? false,
            transform: m.transform as PropertyTransform | undefined,
          });
        }
      }
    }

    return {
      name: data.name as string,
      from_layer,
      from_type,
      to_layer,
      to_type,
      name_template:
        ruleData && typeof ruleData === "object" && "name_template" in ruleData
          ? (ruleData.name_template as string)
          : "",
      property_mappings,
      conditions,
      template_file:
        ruleData && typeof ruleData === "object" && "template" in ruleData
          ? (ruleData.template as string | undefined)
          : undefined,
      create_bidirectional:
        ruleData &&
        typeof ruleData === "object" &&
        "create_bidirectional" in ruleData &&
        ruleData.create_bidirectional !== undefined
          ? (ruleData.create_bidirectional as boolean)
          : true,
    };
  }

  /**
   * Find rules applicable to a source element
   */
  findApplicableRules(source: Element, targetLayer?: string): ProjectionRule[] {
    return this.rules.filter((rule) => {
      // Check layer match
      if (rule.from_layer !== source.layer) return false;

      // Check type match (if specified)
      if (rule.from_type && rule.from_type !== source.type) return false;

      // Check target layer filter (if specified)
      if (targetLayer && rule.to_layer !== targetLayer) return false;

      // Evaluate all conditions (ALL must be true)
      if (rule.conditions) {
        for (const condition of rule.conditions) {
          const evaluator = new ConditionEvaluator(condition);
          if (!evaluator.evaluate(source)) {
            return false;
          }
        }
      }

      return true;
    });
  }

  /**
   * Project a single element to target layer
   */
  async projectElement(
    source: Element,
    targetLayer: string,
    rule?: ProjectionRule,
    dryRun: boolean = false
  ): Promise<Element> {
    // Find rule if not provided
    if (!rule) {
      const applicableRules = this.findApplicableRules(source, targetLayer);
      if (applicableRules.length === 0) {
        throw new Error(
          `No applicable projection rule found for element ${source.id} to layer ${targetLayer}`
        );
      }
      rule = applicableRules[0];
    }

    // Build projected element
    const projected = this.buildProjectedElement(source, rule);

    // If not dry run, add to model
    if (!dryRun) {
      const layer = await this.model.getLayer(rule.to_layer);
      if (!layer) {
        throw new Error(`Target layer ${rule.to_layer} not found`);
      }

      // Add element to layer
      layer.addElement(projected);

      // Register with reference registry if available
      if (this.registry) {
        this.registry.registerElement(projected);
      }
    }

    return projected;
  }

  /**
   * Build projected element from source and rule
   */
  private buildProjectedElement(source: Element, rule: ProjectionRule): Element {
    // Render element name
    const name = this.renderTemplate(rule.name_template, source);

    // Generate element ID
    const id = generateElementId(rule.to_layer, rule.to_type, name);

    // Initialize element data
    const data: Record<string, unknown> = {
      id,
      name,
      type: rule.to_type,
      layer: rule.to_layer,
    };

    // Map properties
    for (const mapping of rule.property_mappings) {
      let value: unknown;

      // Get source value
      if (mapping.source.includes("{") || mapping.source.includes("{{")) {
        // Template string
        value = this.renderTemplate(mapping.source, source);
      } else {
        // Direct property or nested property
        value = this.getSourceValue(source, mapping.source);
      }

      // Use default if null
      if (value === null || value === undefined) {
        if (mapping.required) {
          throw new Error(
            `Required property ${mapping.target} is missing from source ${source.id}`
          );
        }
        value = mapping.default;
      }

      // Apply transformation if specified
      if (mapping.transform && value !== null && value !== undefined) {
        const transformer = new PropertyTransformer(mapping.transform);
        value = transformer.apply(value);
      }

      // Set target property
      if (value !== null && value !== undefined) {
        setNestedProperty(data, mapping.target, value);
      }
    }

    // Add bidirectional reference if enabled
    if (rule.create_bidirectional !== false) {
      if (!data.properties) data.properties = {};
      (data.properties as Record<string, unknown>).realizes = source.id;
    }

    // Create Element instance
    return new Element(data as ConstructorParameters<typeof Element>[0]);
  }

  /**
   * Get value from source element (supports dot notation)
   */
  private getSourceValue(source: Element, path: string): unknown {
    // Check for direct element properties
    if (path === "id") return source.id;
    if (path === "name") return source.name;
    if (path === "type") return source.type;
    if (path === "layer") return source.layer;
    if (path === "description") return source.description;

    // Check nested properties in the properties map
    return getNestedProperty(source.properties, path);
  }

  /**
   * Render template string with source element data
   */
  private renderTemplate(template: string, source: Element): string {
    // Build context
    const context: Record<string, string | Record<string, unknown>> = {
      id: source.id,
      name: source.name,
      type: source.type,
      layer: source.layer ?? "",
      description: source.description ?? "",
      properties: source.properties,
      name_pascal: toPascalCase(source.name),
      name_kebab: toKebabCase(source.name),
      name_snake: toSnakeCase(source.name),
    };

    // Simple template replacement for {source.xxx}
    let result = template;
    const regex = /\{source\.(\w+)\}/g;
    result = result.replace(regex, (match, prop) => {
      return context[prop] !== undefined ? String(context[prop]) : match;
    });

    return result;
  }

  /**
   * Project all elements that match rules
   */
  async projectAll(
    fromLayer?: string,
    toLayer?: string,
    dryRun: boolean = false
  ): Promise<Element[]> {
    const projected: Element[] = [];

    // Get source elements
    const layers = fromLayer
      ? [await this.model.getLayer(fromLayer)]
      : Array.from(this.model.layers.values());

    for (const layer of layers) {
      if (!layer) continue;

      for (const element of layer.listElements()) {
        // Find applicable rules
        const applicableRules = this.findApplicableRules(element, toLayer);

        // Project with each applicable rule
        for (const rule of applicableRules) {
          try {
            const projectedElement = await this.projectElement(
              element,
              rule.to_layer,
              rule,
              dryRun
            );
            projected.push(projectedElement);
          } catch (error) {
            // Log warning and continue
            console.warn(
              `Warning: Failed to project ${element.id} with rule ${rule.name}: ${getErrorMessage(error)}`
            );
          }
        }
      }
    }

    return projected;
  }
}

/**
 * Helper: Get nested property from object
 */
function getNestedProperty(obj: Record<string, unknown>, path: string): unknown {
  const parts = path.split(".");
  let current: unknown = obj;

  for (const part of parts) {
    if (current && typeof current === "object") {
      current = (current as Record<string, unknown>)[part];
    } else {
      return undefined;
    }
  }

  return current;
}

/**
 * Helper: Set nested property on object
 */
function setNestedProperty(obj: Record<string, unknown>, path: string, value: unknown): void {
  const parts = path.split(".");
  let current = obj as Record<string, unknown>;

  for (let i = 0; i < parts.length - 1; i++) {
    const part = parts[i];
    if (!current[part] || typeof current[part] !== "object") {
      current[part] = {};
    }
    current = current[part] as Record<string, unknown>;
  }

  current[parts[parts.length - 1]] = value;
}

/**
 * Helper: Convert to kebab-case
 */
function toKebabCase(text: string): string {
  return text.toLowerCase().replace(/\s+/g, "-").replace(/_/g, "-");
}

/**
 * Helper: Convert to snake_case
 */
function toSnakeCase(text: string): string {
  return text.toLowerCase().replace(/\s+/g, "_").replace(/-/g, "_");
}

/**
 * Helper: Convert to PascalCase
 */
function toPascalCase(text: string): string {
  const words = text.replace(/-/g, " ").replace(/_/g, " ").split(/\s+/);
  return words.map((w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()).join("");
}

/**
 * Helper: Generate element ID
 */
function generateElementId(layer: string, type: string, name: string): string {
  const kebabName = toKebabCase(name);
  return `${layer}.${type}.${kebabName}`;
}
